File: src/tables/hvar.cpp
#include "src/algo.h"
#include "tables.h"

struct HvarMasks
{
    static const int Size = 2;
    static const QString Type;

    static HvarMasks parse(const quint8 *data)
    { return { UInt16::parse(data) }; }

    quint16 innerIndexBits() const { return d & 0x000F; }
    quint16 entrySize() const { return ((d & 0x0030) >> 4) + 1; }

    static QString toString(const HvarMasks &value)
    {
        return QString("Inner index bit count: %1\n"
                       "Map entry size: %2")
            .arg(value.innerIndexBits()).arg(value.entrySize());
    }

    quint16 d;
};

const QString HvarMasks::Type = "Masks";


void parseHvarDeltaSet(Parser &parser)
{
    const auto format = parser.read<HvarMasks>("Entry format");
    const auto count = parser.read<UInt16>("Number of entries");

    const auto innerIndexBits = format.innerIndexBits();
    const auto entrySize = format.entrySize();
    parser.readArray("Entries", count, [&](const auto index){
        if (entrySize == 1) {
            static QHash<quint8, QString> cache;
            const auto entry = parser.peek<UInt8>();
            if (!cache.contains(entry)) {
                const auto outerIndex = entry >> (innerIndexBits + 1);
                const auto innerIndex = entry & ((1 << (innerIndexBits + 1)) - 1);
                const auto value = QString("Outer index: %1\nInner index: %2")
                    .arg(outerIndex).arg(innerIndex);
                cache.insert(entry, value);
            }
            parser.readValue<UInt8>(numberToString(index), cache.value(entry));
        } else if (entrySize == 2) {
            static QHash<quint8, QString> cache;
            const auto entry = parser.peek<UInt16>();
            if (!cache.contains(entry)) {
                const auto outerIndex = entry >> (innerIndexBits + 1);
                const auto innerIndex = entry & ((1 << (innerIndexBits + 1)) - 1);
                const auto value = QString("Outer index: %1\nInner index: %2")
                    .arg(outerIndex).arg(innerIndex);
                cache.insert(entry, value);
            }
            parser.readValue<UInt16>(numberToString(index), cache.value(entry));
        } else {
            throw QString("unsupported entry size");
        }
    });
}

void parseHvar(Parser &parser)
{
    const auto start = parser.offset();

    const auto majorVersion = parser.read<UInt16>("Major version");
    const auto minorVersion = parser.read<UInt16>("Minor version");
    if (!(majorVersion == 1 && minorVersion == 0)) {
        throw QString("invalid table version");
    }

    const auto varStoreOffset = parser.read<Offset32>("Item Variation Store offset");
    const auto advanceWidthMappingOffset = parser.read<OptionalOffset32>("Advance width mapping offset");
    const auto lsbMappingOffset = parser.read<OptionalOffset32>("Left side bearing mapping offset");
    const auto rsbMappingOffset = parser.read<OptionalOffset32>("Right side bearing mapping offset");

    enum class OffsetType {
        VariationStore,
        Advance,
        Lsb,
        Rsb,
    };
    struct Offset {
        OffsetType type;
        quint32 offset;
    };
    std::array<Offset, 4> offsets = {{
        { OffsetType::VariationStore, (quint32)varStoreOffset },
        { OffsetType::Advance, (quint32)advanceWidthMappingOffset },
        { OffsetType::Lsb, (quint32)lsbMappingOffset },
        { OffsetType::Rsb, (quint32)rsbMappingOffset },
    }};

    algo::sort_all_by_key(offsets, &Offset::offset);

    for (const auto offset : offsets) {
        if (offset.offset == 0) {
            continue;
        }

        parser.advanceTo(start + offset.offset);
        switch (offset.type) {
        case OffsetType::VariationStore: {
            parser.beginGroup("Item Variation Store");
            parseItemVariationStore(parser);
            parser.endGroup();
            break;
        }
        case OffsetType::Advance: {
            parser.beginGroup("Advance Width Mapping");
            parseHvarDeltaSet(parser);
            parser.endGroup();
            break;
        }
        case OffsetType::Lsb: {
            parser.beginGroup("Left Side Bearing Mapping");
            parseHvarDeltaSet(parser);
            parser.endGroup();
            break;
        }
        case OffsetType::Rsb: {
            parser.beginGroup("Right Side Bearing Mapping");
            parseHvarDeltaSet(parser);
            parser.endGroup();
            break;
        }
        }
    }
}

File: core/src/tables/hvar.rs
use crate::parser::*;
use crate::{ValueType, Error, Result};

#[derive(Clone, Copy, Debug)]
struct Masks(u16);

impl Masks {
    #[inline] fn inner_index_bits(self) -> u16 { self.0 & 0x000F }
    #[inline] fn entry_size(self) -> u16 { ((self.0 & 0x0030) >> 4) + 1 }
}

impl FromData for Masks {
    const NAME: ValueType = ValueType::Masks;

    #[inline]
    fn parse(data: &[u8]) -> Result<Self> {
        u16::parse(data).map(Masks)
    }
}

impl std::fmt::Display for Masks {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Inner index bit count: {}\nMap entry size: {}",
               self.inner_index_bits(), self.entry_size())
    }
}

pub fn parse(parser: &mut Parser) -> Result<()> {
    let start = parser.offset();

    let major_version: u16 = parser.read("Major version")?;
    let minor_version: u16 = parser.read("Minor version")?;
    if !(major_version == 1 && minor_version == 0) {
        return Err(Error::InvalidTableVersion);
    }

    let var_store_offset = parser.read::<Offset32>("Item variation store offset")?.to_usize();
    let advance_width_mapping_offset = parser.read::<OptionalOffset32>("Advance width mapping offset")?.to_usize();
    let lsb_mapping_offset = parser.read::<OptionalOffset32>("Left side bearing mapping offset")?.to_usize();
    let rsb_mapping_offset = parser.read::<OptionalOffset32>("Right side bearing mapping offset")?.to_usize();

    parser.jump_to(start + var_store_offset)?;
    parser.begin_group("Item variation store");
    super::mvar::parse_item_variation_store(parser)?;
    parser.end_group();

    if let Some(offset) = advance_width_mapping_offset {
        parser.jump_to(start + offset)?;
        parser.begin_group("Advance width mapping");
        parse_delta_set(parser)?;
        parser.end_group();
    }

    if let Some(offset) = lsb_mapping_offset {
        parser.jump_to(start + offset)?;
        parser.begin_group("Left side bearing mapping");
        parse_delta_set(parser)?;
        parser.end_group();
    }

    if let Some(offset) = rsb_mapping_offset {
        parser.jump_to(start + offset)?;
        parser.begin_group("Right side bearing mapping");
        parse_delta_set(parser)?;
        parser.end_group();
    }

    Ok(())
}

pub fn parse_delta_set(parser: &mut Parser) -> Result<()> {
    let format = parser.read::<Masks>("Entry format")?;
    let count = parser.read::<u16>("Number of entries")?;

    let inner_index_bits = format.inner_index_bits();
    let entry_size = format.entry_size();
    parser.begin_group("Entries");
    for _ in 0..count {
        if entry_size == 1 {
            let entry = parser.peek::<u8>()?;
            let outer_index = entry >> (inner_index_bits + 1);
            let inner_index = entry & ((1 << (inner_index_bits + 1)) - 1);
            parser.read_value(1, "Entry",
                format!("Outer index: {}\nInner index: {}", outer_index, inner_index))?;
        } else if entry_size == 2 {
            let entry = parser.peek::<u16>()?;
            let outer_index = entry >> (inner_index_bits + 1);
            let inner_index = entry & ((1 << (inner_index_bits + 1)) - 1);
            parser.read_value(2, "Entry",
                 format!("Outer index: {}\nInner index: {}", outer_index, inner_index))?;
        } else {
            return Err(Error::InvalidValue);
        }
    }
    parser.end_group();

    Ok(())
}
